libbeat/autodiscover/providers/kubernetes/pod.go

```diff
@@ -392,6 +391,10 @@ func (p *pod) emitEvents(pod *kubernetes.Pod, flag string, containers []kubernet
 			},
 		}
 		p.publish(event)
+	}
 
+	// Publish the container level hints finally.
+	for _, event := range events {
+		p.publish(event)
```

Comment:
Why does the order matter if the pod event is only going to be sent if there was some container level event?

Reply:
if a pod has port 8080 exposed on container a then the container level meta would be:

```
pod: p
container.name: a
```

the pod level meta would be:
```
pod: p #no container meta
```

the config for both would just be:
```
module:
hosts: ["host:port"]
```

when autodiscover sees both events, it will create a single config but the meta keeps getting updated each time a new event comes in. if the pod event comes last then it will just remove the meta that has container meta on it as per:
https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L216

To avoid losing container meta, we ensure that pod level hint is sent first so that it can be overridden by a container event that subsequently matches the same config.

Reply:
But we only send the pod event if `len(events) != 0`, so why does it matter if we send it before of after the containers loop?

Reply:
Oh, sorry, I was thinking on the opposite condition :facepalm: All good :+1: